python/lsst/ts/m1m3/utils/thermocouples.py

```diff
@@ -52,318 +55,1000 @@ async def get_scanner_data(
+                                    == "M"  # last char “M”  → z = 0
+                                    else (
+                                        1
+                                        if the_thermocouple.name[-1]
+                                        == "F"  # last char “F”  → z = 0.5
+                                        else 0
+                                    )
+                                ),
+                            )
+                        }
+                    )
                 else:
-                    chunk_data.rename(columns={name: tc.name}, inplace=True)
+                    coord.update(
+                        {
+                            the_thermocouple.name: (
+                                the_thermocouple.x_position,
+                                the_thermocouple.y_position,
+                            )
+                        }
+                    )
+        if make_3d_map:
+            common = [c for c in data.columns if c in coord]
+            xyz = np.array([coord[n] for n in common])
+        else:
+            common = [c for c in data.columns if c + "F" in coord]
+            xyz = np.array([coord[n + "F"] for n in common])
 
-            chunk_data.drop(columns=to_drop, inplace=True)
+        return xyz.T, data[common]
 
-            ret[chunk_index] = chunk_data.resample(freq).median()
+    def calculate_vertical_differences(self) -> None:
+        """Calculate vertical gradients."""
+        # Suffixes
+        b_suffixes = ["B", "B1"]
+        f_suffix = "F"
 
-        return ret
+        # Store results
+        result = {}
 
-    tasks: list[asyncio.Task] = []
+        if self.all_thermocouples_dataframe is not None:
 
-    async with asyncio.TaskGroup() as tg:
-        # Loop through every thermal scanner
-        for scanner in Scanner:
-            tasks.append(tg.create_task(load_chunk(scanner)))
+            for col in self.all_thermocouples_dataframe.columns:
+                for b_suffix in b_suffixes:
+                    if col.endswith(b_suffix):
+                        base_name = col[: -len(b_suffix)]
+                        f_col = base_name + f_suffix
+                        if f_col in self.all_thermocouples_dataframe.columns:
+                            result[base_name] = (
+                                self.all_thermocouples_dataframe[f_col]
+                                - self.all_thermocouples_dataframe[col]
+                            )
 
-    scanner_dataframe = pd.DataFrame()
+            # Create the new DataFrame
+            self.vertical_cell_gradient_dataframe = pd.DataFrame(result)
 
-    for scanner_data in [t.result() for t in tasks]:
-        if scanner_data is None:
-            continue
+    def calculate_gradients_xy_r(
+        self,
+        remove_nonstandard_cells: bool = True,
+        use_3d_dataset: bool = True,
+    ) -> pd.DataFrame:
+        """
+        Estimate per-point (gx, gy) and radial gradient gr
```

Comment:
Also worth updating this docstring since we also optionally (but by default) calculate gz